src/moscot/backends/ott/output.py

```diff
@@ -178,30 +229,792 @@ def to(self, device: Optional[Device_t] = None) -> "OTTOutput":  # noqa: D102
+        if isinstance(device, str) and ":" in device:
+            device, ix = device.split(":")
+            idx = int(ix)
+        else:
+            idx = 0
+
+        if not isinstance(device, xla_ext.Device):
+            try:
+                device = jax.devices(device)[idx]
+            except IndexError:
+                raise IndexError(f"Unable to fetch the device with `id={idx}`.")
+
+        out = jax.device_put(self._output, device)
+        return CondNeuralDualOutput(output=out, training_logs=self.training_logs)
+
+    def push(self, cond: ArrayLike, x: ArrayLike) -> ArrayLike:  # type: ignore[override]
+        """Push distribution `x` conditioned on condition `cond`.
+
+        Parameters
+        ----------
+        cond
+            Condition of conditional neural OT.
+        x
+            Distribution to push.
+
+        Returns
+        -------
+        Pushed distribution.
+        """
+        if x.ndim not in (1, 2):
+            raise ValueError(f"Expected 1D or 2D array, found `{x.ndim}`.")
+        return self._apply(cond, x, forward=True)
+
+    def pull(self, cond: ArrayLike, x: ArrayLike) -> ArrayLike:  # type: ignore[override]
+        """Pull distribution `x` conditioned on condition `cond`.
+
+        Parameters
+        ----------
+        cond
+            Condition of conditional neural OT.
+        x
+            Distribution to pull.
+
+        Returns
+        -------
+        Pulled distribution.
+        """
+        if x.ndim not in (1, 2):
+            raise ValueError(f"Expected 1D or 2D array, found `{x.ndim}`.")
+        return self._apply(cond, x, forward=False)
+
+    def evaluate_f(self, cond: ArrayLike, x: ArrayLike) -> ArrayLike:  # type:ignore[override]
+        """Apply forward potential to `x` conditionend on condition `cond`.
+
+        Parameters
+        ----------
+        cond
+            Condition of conditional neural OT.
+        x
+            Distribution to apply potential to.
+
+        Returns
+        -------
+        Forward potential evaluated at `x`.
+        """
+        if x.ndim not in (1, 2):
+            raise ValueError(f"Expected 1D or 2D array, found `{x.ndim}`.")
+        return jax.vmap(self._output.f)(cond, x)
+
+    def evaluate_g(self, cond: ArrayLike, x: ArrayLike) -> ArrayLike:  # type:ignore[override]
+        """Apply backward potential to `x` conditionend on condition `cond`.
+
+        Parameters
+        ----------
+        cond
+            Condition of conditional neural OT.
+        x
+            Distribution to apply backward potential to.
+
+        Returns
+        -------
+        Backward potential evaluated at `x`.
+        """
+        if x.ndim not in (1, 2):
+            raise ValueError(f"Expected 1D or 2D array, found `{x.ndim}`.")
+        return jax.vmap(self._output.g)(cond, x)
+
+
+class GapNeuralOutput(ConvergencePlotterMixin, BaseNeuralOutput):
```

Comment:
I'd consider having a general `FunctionalOutput`, which needs an object that implements, e.g., `transport()` and `distance()` functions (this can be done via a class with abstract methods; in the case of NeuralDual, I'd subclass this class + ott's NeuralDual, which will then satisfy the interface.